src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,355 @@
+import org.prebid.server.analytics.reporter.greenbids.model.MediaTypes;
+import org.prebid.server.auction.model.AuctionContext;
+import org.prebid.server.auction.model.BidRejectionTracker;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.EncodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.log.Logger;
+import org.prebid.server.log.LoggerFactory;
+import org.prebid.server.proto.openrtb.ext.request.ExtImpPrebid;
+import org.prebid.server.proto.openrtb.ext.request.ExtRequest;
+import org.prebid.server.proto.openrtb.ext.request.ExtRequestPrebid;
+import org.prebid.server.proto.openrtb.ext.request.ExtStoredRequest;
+import org.prebid.server.proto.openrtb.ext.response.seatnonbid.NonBid;
+import org.prebid.server.proto.openrtb.ext.response.seatnonbid.SeatNonBid;
+import org.prebid.server.util.HttpUtil;
+import org.prebid.server.vertx.httpclient.HttpClient;
+import org.prebid.server.vertx.httpclient.model.HttpClientResponse;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GreenbidsAnalyticsReporter implements AnalyticsReporter {
+
+    private static final int RANGE_16_BIT_INTEGER_DIVISION_BASIS = 0x10000;
+    private static final String PREBID_SERVER_VERSION = "3.0.0+server";
+    public static final Logger logger = LoggerFactory.getLogger(GreenbidsAnalyticsReporter.class);
+    private final GreenbidsAnalyticsProperties greenbidsAnalyticsProperties;
+    private final JacksonMapper jacksonMapper;
+    private final HttpClient httpClient;
+
+    public GreenbidsAnalyticsReporter(
+            GreenbidsAnalyticsProperties greenbidsAnalyticsProperties,
+            JacksonMapper jacksonMapper,
+            HttpClient httpClient) {
+        this.greenbidsAnalyticsProperties = Objects.requireNonNull(greenbidsAnalyticsProperties);
+        this.jacksonMapper = Objects.requireNonNull(jacksonMapper);
+        this.httpClient = Objects.requireNonNull(httpClient);
+    }
+
+    @Override
+    public <T> Future<Void> processEvent(T event) {
+        final AuctionContext auctionContext;
+        final BidResponse bidResponse;
+
+        if (event instanceof AmpEvent ampEvent) {
+            auctionContext = ampEvent.getAuctionContext();
+            bidResponse = ampEvent.getBidResponse();
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            auctionContext = auctionEvent.getAuctionContext();
+            bidResponse = auctionEvent.getBidResponse();
+        } else {
+            auctionContext = null;
+            bidResponse = null;
+        }
+
+        if (bidResponse == null || auctionContext == null) {
+            return Future.failedFuture("Bid response or auction context cannot be null");
```

Comment:
returning the future of exception is better practice

Reply:
you can see that the exception are handled higher in the calling stack and the corresponding metric will be logged, so please use one of the expected (depending on the failure) or the common one `PrebidException`
```
    private <T> Future<Void> processFail(Throwable exception, T event, String reporterName) {
        final MetricName failedResult;
        if (exception instanceof TimeoutException || exception instanceof ConnectTimeoutException) {
            failedResult = MetricName.timeout;
        } else if (exception instanceof InvalidRequestException) {
            failedResult = MetricName.badinput;
        } else {
            failedResult = MetricName.err;
        }
        updateMetricsByEventType(event, reporterName, failedResult);
        return Future.failedFuture(exception);
    }
 ```